### PR TITLE
Alejandro/whackamole

### PR DESCRIPTION
**Pull request by Sl3v7n**

### CODE REVIEW

File: src/main/java/org/minions/devfund/alejandro/WhackAMole.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    /**
+     * This method will fill the matrix with moles.
+     * @param size the size of the matrix.
+     */
+    private void fillMoles(int size) {
+
+        for (int i = 0; i < size; i++) {
+            boolean p = false;
+            while (!p) {
+                int randomX = (int) (Math.random() * size + 0);
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/WhackAMole.java]: MINOR Use "java.util.Random.nextInt()" instead. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2140)

File: src/main/java/org/minions/devfund/alejandro/WhackAMole.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                int randomY = (int) (Math.random() * size + 0);
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/WhackAMole.java]: MINOR Use "java.util.Random.nextInt()" instead. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2140)

File: src/main/java/org/minions/devfund/alejandro/WhackAMole.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+ */
+public class WhackAMole {
+
+    private int score;
+    private int molesLeft;
+    private int attemptsLeft;
+    private int moles;
+    private char[][] moleGrid;
+    private int size;
+    private final char mole = 'M';
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/WhackAMole.java]: MINOR Make this final field static too. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1170)

File: src/main/java/org/minions/devfund/alejandro/WhackAMole.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    private final char nothing = '*';
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/WhackAMole.java]: MINOR Make this final field static too. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1170)

File: src/main/java/org/minions/devfund/alejandro/MainWhackAMole.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
+package org.minions.devfund.alejandro;
+import java.util.Scanner;
+
+/**
+ * class for main wack a mole game.
+ */
+public abstract class  MainWhackAMole {
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/MainWhackAMole.java]: MINOR Convert the abstract class "MainWhackAMole" into an interface. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1610)

**Comment by carledriss** [comment]
SonarQube analysis reported 5 issues
* ![MINOR][MINOR] 5 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/org/minions/devfund/alejandro/WhackAMole.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+ */
+public class WhackAMole {
+
+    private int score;
+    private int molesLeft;
+    private int moles;
+    private int attemptsLeft;
+    private final char[][] moleGrid;
+    private final int size;
+    private final char mole = 'M';
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/WhackAMole.java]: MINOR Make this final field static too. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1170)

File: src/main/java/org/minions/devfund/alejandro/WhackAMole.java [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    private final char nothing = '*';
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/WhackAMole.java]: MINOR Make this final field static too. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1170)

File: src/main/java/org/minions/devfund/alejandro/MainWhackAMole.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+/**
+ * class for main wack a mole game.
+ */
+public interface MainWhackAMole {
+
+    /**
+     * This is the main method to start running the game.
+     * @param args arguments.
+     */
+    static void main(final String[] args) {
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/MainWhackAMole.java]: MAJOR Remove this unused method parameter "args". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1172)

**Comment by carledriss** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/org/minions/devfund/alejandro/WhackAMole.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+     * Given a location, place a mole at that location.
+     * @param x location.
+     * @param y location.
+     * @return true if you can. (Also update number of moles left.)
+     */
+    boolean place(int x, int y) {
+        if (this.moleGrid[x][y] != 'M') {
+            this.moleGrid[x][y] = 'M';
+            this.molesLeft++;
+            System.out.println(molesLeft);
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/WhackAMole.java]: Remove this sout

File: src/main/java/org/minions/devfund/alejandro/WhackAMole.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+     */
+    void whack(int x, int y) {
+        if (this.moleGrid[x][y] == 'M') {
+            this.score++;
+            this.attemptsLeft--;
+            molesLeft--;
+        } else {
+            this.attemptsLeft--;
+        }
+        System.out.println(molesLeft);
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/WhackAMole.java]: Remove this sout

File: src/main/java/org/minions/devfund/alejandro/WhackAMole.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    /**
+     * Print the grid without showing where the moles are.
+     * For every spot that has recorded a “whacked mole,” print out a W, or * otherwise.
+     */
+    void printGridToUser() {
+        for (char[] aMoleGrid : moleGrid) {
+            for (char[] aMoleGrid1 : moleGrid) {
+                System.out.print('*');
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/WhackAMole.java]: Consider using a StringBuilder here
in order to make it testable.

File: src/main/java/org/minions/devfund/alejandro/WhackAMole.java [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+    /**
+     * Print the grid completely. This is effectively dumping the 2d array on the screen.
+     * The places where the moles are should be printed as an ‘M’.
+     * The places where the moles have been whacked should be printed as a ‘W’.
+     * The places that don’t have a mole should be printed as *.
+     */
+    void printGrid() {
+        for (char[] aMoleGrid : moleGrid) {
+            for (int y = 0; y < moleGrid.length; y++) {
+                System.out.print(aMoleGrid[y]);
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/WhackAMole.java]: Consider using a StringBuilder here
in order to make it testable.

File: src/main/java/org/minions/devfund/alejandro/WhackAMole.java [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+    /**
+     * Print the grid without showing where the moles are.
+     * For every spot that has recorded a “whacked mole,” print out a W, or * otherwise.
+     */
+    void printGridToUser() {
+
+        this.auxialiryBuilder.setLength(0);
+        for (char[] aMoleGrid : this.moleGrid) {
+            for (char[] aMoleGrid1 : this.moleGrid) {
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/WhackAMole.java]: MINOR Remove this unused "aMoleGrid1" local variable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1481)

File: src/main/java/org/minions/devfund/alejandro/WhackAMole.java [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    /**
+     * Print the grid without showing where the moles are.
+     * For every spot that has recorded a “whacked mole,” print out a W, or * otherwise.
+     */
+    void printGridToUser() {
+
+        this.auxialiryBuilder.setLength(0);
+        for (char[] aMoleGrid : this.moleGrid) {
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/WhackAMole.java]: MINOR Remove this unused "aMoleGrid" local variable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1481)

**Comment by carledriss** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by Sl3v7n** [comment]
image

**Comment by Sl3v7n** [comment]
Execution in Edx platform

> 
[Executed at: Sun Apr 15 18:33:38 PDT 2018]

Executing command: javac -J-Xmx256m -d . WhackAMole.java (Passed)

[Executed at: Sun Apr 15 18:33:48 PDT 2018]

All tests passed
Executing command: javac -J-Xmx256m -d . WhackAMole.java (Passed)
Executing command: javac -J-Xmx256m -d . -cp .:/home/vlibs/java/voc-grader.jar:/home/vlibs/java/junit-4.12.jar:/home/vlibs/java/hamcrest-core-1.3.jar /home/ccc_v1_c79431__48717/asn12900_Whack_a_Mole/asn12901_JUnit/asnlib/WhackAMoleTestGrader.java (Passed)
Running: testConstructorGrid
    Finished: testConstructorGrid
Running: testConstructorNumOfAttempts
    Finished: testConstructorNumOfAttempts
Running: testPlaceFail
1
    Finished: testPlaceFail
Running: testWhackSucceedAttempts
1
0
    Finished: testWhackSucceedAttempts
Running: testWhackFailAttempts
1
1
    Finished: testWhackFailAttempts
Running: testPlaceSucceed
1
2
    Finished: testPlaceSucceed
Running: testWhackFailMolesLeft
1
1
    Finished: testWhackFailMolesLeft
Running: testWhackSucceedMolesLeft
1
0
    Finished: testWhackSucceedMolesLeft
Tests run: 8, Passed: 8, Failed: 0
Score: 40/40

File: src/main/java/org/minions/devfund/alejandro/WhackAMole.java [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+     * @param x location.
+     * @param y location.
+     */
+    void whack(int x, int y) {
+
+        if (this.moleGrid[x][y] == 'M') {
+            this.score++;
+            this.attemptsLeft--;
+            molesLeft--;
+        } else {
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/WhackAMole.java]: Remove this unnecessary else condition
you can directly declare this.attemptsLeft--; and remove it from if conditional

**Comment by carledriss** [comment]
@Sl3v7n 
Don't forget to add unit tests

File: src/main/java/org/minions/devfund/alejandro/WhackAMole.java [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+    /**
+     * Print the grid without showing where the moles are.
+     * For every spot that has recorded a “whacked mole,” print out a W, or * otherwise.
+     * @return an string.
+     */
+    public String printGridToUser() {
+
+        this.auxiliaryBuilder.setLength(0);
+        for (char[] aMoleGrid : this.moleGrid) {
+            for (char[] aMoleGrid1 : this.moleGrid) {
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/WhackAMole.java]: MINOR Remove this unused "aMoleGrid1" local variable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1481)

File: src/main/java/org/minions/devfund/alejandro/WhackAMole.java [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+
+    /**
+     * Print the grid without showing where the moles are.
+     * For every spot that has recorded a “whacked mole,” print out a W, or * otherwise.
+     * @return an string.
+     */
+    public String printGridToUser() {
+
+        this.auxiliaryBuilder.setLength(0);
+        for (char[] aMoleGrid : this.moleGrid) {
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/WhackAMole.java]: MINOR Remove this unused "aMoleGrid" local variable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1481)

File: src/test/java/org/minions/devfund/alejandro/tests/whackamole/WhackAMoleTest.java [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+ * This class contain the test cases for WhackAMole class.
+ */
+
+public class WhackAMoleTest {
+
+    private final int numAttempts = 10;
+    private final int gridDimension = 4;
+    private int xPosition = 2;
+    private int yPosition = 2;
+    private final String expectedChar = "****\n****\n****\n****\n";
```
Comment (carledriss) [inline src/test/java/org/minions/devfund/alejandro/tests/whackamole/WhackAMoleTest.java]: MINOR Make this final field static too. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1170)

File: src/test/java/org/minions/devfund/alejandro/tests/whackamole/WhackAMoleTest.java [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+    private final String expectedCharMoles = "****\n****\n**M*\n****\n";
```
Comment (carledriss) [inline src/test/java/org/minions/devfund/alejandro/tests/whackamole/WhackAMoleTest.java]: MINOR Make this final field static too. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1170)

File: src/test/java/org/minions/devfund/alejandro/tests/whackamole/WhackAMoleTest.java [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+    private final String expectedCharWhacks = "****\n****\n**W*\n****\n";
```
Comment (carledriss) [inline src/test/java/org/minions/devfund/alejandro/tests/whackamole/WhackAMoleTest.java]: MINOR Make this final field static too. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1170)

File: src/test/java/org/minions/devfund/alejandro/tests/whackamole/WhackAMoleTest.java [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * This class contain the test cases for WhackAMole class.
+ */
+
+public class WhackAMoleTest {
+
+    private final int numAttempts = 10;
```
Comment (carledriss) [inline src/test/java/org/minions/devfund/alejandro/tests/whackamole/WhackAMoleTest.java]: MINOR Make this final field static too. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1170)

File: src/test/java/org/minions/devfund/alejandro/tests/whackamole/WhackAMoleTest.java [hunk 22]
```diff
@@ -0,0 +1,153 @@
...
+    private final int gridDimension = 4;
```
Comment (carledriss) [inline src/test/java/org/minions/devfund/alejandro/tests/whackamole/WhackAMoleTest.java]: MINOR Make this final field static too. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1170)

**Comment by carledriss** [comment]
SonarQube analysis reported 7 issues
* ![MINOR][MINOR] 7 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/org/minions/devfund/alejandro/WhackAMole.java [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+ * It contains three integer instance variables called score, molesLeft, and attemptsLeft.
+ * Make one more instance variable called moleGrid which is a 2-dimensional array of chars.
+ */
+public class WhackAMole {
+
+    private int score;
+    private int molesLeft;
+    private int attemptsLeft;
+    private char[][] moleGrid;
+    private StringBuilder auxiliaryBuilder = new StringBuilder();
```
Comment (carledriss) [inline src/main/java/org/minions/devfund/alejandro/WhackAMole.java]: Declare as attribute of class will have issues in the future.
Please declare as local variable inside method.

File: src/test/java/org/minions/devfund/alejandro/tests/whackamole/WhackAMoleTest.java [hunk 24]
```diff
@@ -0,0 +1,153 @@
...
+public class WhackAMoleTest {
+
+    private static final int NUM_ATTEMPTS = 10;
+    private static final int GRID_DIMENSION = 4;
+    private static int xPosition = 2;
+    private static int yPosition = 2;
+    private static final String EXPECTED_CHAR = "****\n****\n****\n****\n";
+    private static final String EXPECTED_CHAR_MOLES = "****\n****\n**M*\n****\n";
+    private static final String EXPECTED_CHAR_WHACKS = "****\n****\n**W*\n****\n";
+    private WhackAMole whackAMole = new WhackAMole(NUM_ATTEMPTS, GRID_DIMENSION);
```
Comment (carledriss) [inline src/test/java/org/minions/devfund/alejandro/tests/whackamole/WhackAMoleTest.java]: Please move this inside 2@Before setUp() method.

**Comment by carledriss** [comment]
LGTM! 👍

**Pull request merged**